libs/ledger/src/chain/main_chain.cpp

```diff
@@ -403,47 +529,82 @@ bool MainChain::RemoveBlock(BlockHash const &hash)
 /**
  * Walk the block history starting from the heaviest block
  *
- * @param limit The maximum number of blocks to be returned
+ * @param lowest_block_number The minimum allowed block_number to be returned
  * @return The array of blocks
  * @throws std::runtime_error if a block lookup occurs
  */
-MainChain::Blocks MainChain::GetHeaviestChain(uint64_t limit) const
+MainChain::Blocks MainChain::GetHeaviestChain(uint64_t lowest_block_number) const
 {
   // Note: min needs a reference to something, so this is a workaround since UPPER_BOUND is a
   // constexpr
-  limit = std::min(limit, uint64_t{MainChain::UPPER_BOUND});
   MilliTimer myTimer("MainChain::HeaviestChain");
 
   FETCH_LOCK(lock_);
 
-  return GetChainPreceding(GetHeaviestBlockHash(), limit);
+  return GetChainPreceding(GetHeaviestBlockHash(), lowest_block_number);
+}
+
+MainChain::IntBlockPtr MainChain::GetLabeledSubchainStart() const
+{
+  return labeled_subchain_start_;
+}
+
+/**
+ * Update blocks of the current heaviest chain setting their chain_label
+ * equal to heaviest_.ChainLabel().
+ *
+ * @param limit the earliest block number, this colouring stops at.
+ */
+void MainChain::ColourHeaviestChainBlocks(uint64_t limit) const
+{
+  MilliTimer myTimer("MainChain::ColourHeaviestChainBlocks");
+  FETCH_LOCK(lock_);
+  assert(heaviest_.ChainLabel() != 0);
+
+  auto block = GetLabeledSubchainStart();
+  assert(block);
+  while (block->block_number > limit)
+  {
+    assert(!block->IsGenesis());
+    if (!LookupBlock(block->previous_hash, block))
+    {
+      throw std::runtime_error("Cannot find a block for hash");
+    }
+    block->chain_label = heaviest_.ChainLabel();
+  }
+  labeled_subchain_start_ = block;
 }
 
 /**
  * Walk the block history collecting blocks until either genesis or the block limit is reached
  *
  * @param start The hash of the first block
- * @param limit The maximum number of blocks to be returned
+ * @param lowest_block_number The minimum allowed block_number to be returned
  * @return The array of blocks
  * @throws std::runtime_error if a block lookup occurs
  */
-MainChain::Blocks MainChain::GetChainPreceding(BlockHash start, uint64_t limit) const
+MainChain::Blocks MainChain::GetChainPreceding(BlockHash start, uint64_t lowest_block_number) const
 {
-  if (limit == 0)
-  {
-    return Blocks{};
-  }
-  limit = std::min(limit, static_cast<uint64_t>(MainChain::UPPER_BOUND));
   MilliTimer myTimer("MainChain::ChainPreceding");
 
   FETCH_LOCK(lock_);
 
+  // asserting genesis block has a number of 0, and everything else is above
+  assert(GetBlock(chain::GENESIS_DIGEST));
+  assert(GetBlock(chain::GENESIS_DIGEST)->block_number == 0);
+#ifndef NDEBUG
+  for (auto const &block_entry : block_chain_)
+  {
+    assert(block_entry.second->block_number > 0 || block_entry.first == chain::GENESIS_DIGEST);
```

Comment:
Is this NDEBUG neccessary? Assertions depend on NDEBUG anyway - I believe in Release mode we do not define it, whereas in Debug mode we do

Reply:
> Is this NDEBUG neccessary?
Not at all, provided we're sure that the empty loop will be optimized out in Release mode.